File: baseLibrary/commonsdk/src/main/java/com/common/recorder/MyMediaRecorder.java
package com.common.recorder;

import android.media.MediaRecorder;

import com.common.log.MyLog;

import java.io.IOException;

public class MyMediaRecorder {
    public final static String TAG = "MyMediaRecorder";

    int audioSource = MediaRecorder.AudioSource.MIC;
    int outputFormat = MediaRecorder.OutputFormat.MPEG_4;
    int audioEncoder = MediaRecorder.AudioEncoder.AAC;
    int audioChannel = 1;
    int audioSamplingRate = 44100;
    int audioEncodingBitRate = 192000;
    long mStartRecordingTs = 0;
    int mDuration = 0;
    MediaRecorder mMediaRecorder;

//    String filePath;

    public int getAudioSource() {
        return audioSource;
    }

    public void setAudioSource(int audioSource) {
        this.audioSource = audioSource;
    }

    public int getOutputFormat() {
        return outputFormat;
    }

    public void setOutputFormat(int outputFormat) {
        this.outputFormat = outputFormat;
    }

    public int getAudioEncoder() {
        return audioEncoder;
    }

    public void setAudioEncoder(int audioEncoder) {
        this.audioEncoder = audioEncoder;
    }

    public int getAudioChannel() {
        return audioChannel;
    }

    public void setAudioChannel(int audioChannel) {
        this.audioChannel = audioChannel;
    }

    public int getAudioSamplingRate() {
        return audioSamplingRate;
    }

    public void setAudioSamplingRate(int audioSamplingRate) {
        this.audioSamplingRate = audioSamplingRate;
    }

    public int getAudioEncodingBitRate() {
        return audioEncodingBitRate;
    }

    public void setAudioEncodingBitRate(int audioEncodingBitRate) {
        this.audioEncodingBitRate = audioEncodingBitRate;
    }

    private void config(String filePath) {
        if (mMediaRecorder == null) {
            mMediaRecorder = new MediaRecorder();
        }
        mMediaRecorder.setAudioSource(audioSource);
        mMediaRecorder.setOutputFormat(outputFormat);
        mMediaRecorder.setOutputFile(filePath);
        mMediaRecorder.setAudioEncoder(audioEncoder);
        mMediaRecorder.setAudioChannels(audioChannel);
        mMediaRecorder.setAudioSamplingRate(audioSamplingRate);
        mMediaRecorder.setAudioEncodingBitRate(audioEncodingBitRate);
    }

    public void start(String filePath) {
        config(filePath);
        try {
            if (mMediaRecorder != null) {
                mMediaRecorder.prepare();
                mMediaRecorder.start();
            }
            mStartRecordingTs = System.currentTimeMillis();
        } catch (IOException e) {
            MyLog.e(TAG, e);
        }
    }

    public void stop() {
        if (mMediaRecorder != null) {
            mMediaRecorder.reset();
        }
        mDuration = (int) (System.currentTimeMillis() - mStartRecordingTs);
    }

    public void destroy() {
        if (mMediaRecorder != null) {
            mMediaRecorder.reset();
            mMediaRecorder.release();
            mMediaRecorder = null;
        }
    }

    /**
     * 单位ms
     *
     * @return
     */
    public int getDuration() {
        return mDuration;
    }

    public static Builder newBuilder() {
        return new Builder();
    }

    public static class Builder {
        MyMediaRecorder mParams = new MyMediaRecorder();

        Builder() {
        }

        public Builder setAudioSource(int audioSource) {
            mParams.setAudioSource(audioSource);
            return this;
        }

        public Builder setOutputFormat(int outputFormat) {
            mParams.setOutputFormat(outputFormat);
            return this;
        }

        public Builder setAudioEncoder(int audioEncoder) {
            mParams.setAudioEncoder(audioEncoder);
            return this;
        }

        public Builder setAudioChannel(int audioChannel) {
            mParams.setAudioChannel(audioChannel);
            return this;
        }

        public Builder setAudioSamplingRate(int audioSamplingRate) {
            mParams.setAudioSamplingRate(audioSamplingRate);
            return this;
        }

        public Builder setAudioEncodingBitRate(int audioEncodingBitRate) {
            mParams.setAudioEncodingBitRate(audioEncodingBitRate);
            return this;
        }

        public MyMediaRecorder build() {
            return mParams;
        }

    }
}

File: baseLibrary/commonsdk/src/main/java/com/common/recorder/MyMediaRecorder.kt
package com.common.recorder

import android.media.MediaRecorder
import android.os.Handler
import android.os.Looper
import android.os.Message

import com.common.log.MyLog

class MyMediaRecorder {

    //    String filePath;
    val MSG_UPDATE_VOLUME = 81

    var audioSource = MediaRecorder.AudioSource.MIC
    var outputFormat = MediaRecorder.OutputFormat.MPEG_4
    var audioEncoder = MediaRecorder.AudioEncoder.AAC
    var audioChannel = 1
    var audioSamplingRate = 44100
    var audioEncodingBitRate = 192000
    private var mStartRecordingTs: Long = 0
    /**
     * 单位ms
     *
     * @return
     */
    var mDuration = 0
        internal set
    private var mMediaRecorder: MediaRecorder? = null
    private var mRecording = false
    private var mUiHandler: Handler? = null
    private var mVolumeListener: ((Int) -> Unit)? = null

    private fun config(filePath: String) {
        if (mMediaRecorder == null) {
            mMediaRecorder = MediaRecorder()
        }
        mMediaRecorder?.setAudioSource(audioSource)
        mMediaRecorder?.setOutputFormat(outputFormat)
        mMediaRecorder?.setOutputFile(filePath)
        mMediaRecorder?.setAudioEncoder(audioEncoder)
        mMediaRecorder?.setAudioChannels(audioChannel)
        mMediaRecorder?.setAudioSamplingRate(audioSamplingRate)
        mMediaRecorder?.setAudioEncodingBitRate(audioEncodingBitRate)
    }

    fun start(filePath: String, callback: ((Int) -> Unit)?) {
        config(filePath)
        try {
            if (mRecording) {
                mMediaRecorder?.reset()
            }
            mMediaRecorder?.prepare()
            mMediaRecorder?.start()
            mDuration = 0
            mStartRecordingTs = System.currentTimeMillis()
            mRecording = true
            mVolumeListener = callback
            if (callback != null) {
                updateVolume()
            }
        } catch (e: Exception) {
            MyLog.e(TAG, e)
        }
    }

    fun stop() {
        if (mRecording) {
            mDuration = (System.currentTimeMillis() - mStartRecordingTs).toInt()
        }
        mRecording = false
        mMediaRecorder?.reset()
        mUiHandler?.removeCallbacksAndMessages(null)
        mVolumeListener = null
    }

    fun destroy() {
        mRecording = false
        mMediaRecorder?.reset()
        mMediaRecorder?.release()
        mMediaRecorder = null
        mUiHandler?.removeCallbacksAndMessages(null)
        mUiHandler = null
        mVolumeListener = null
    }

    private fun updateVolume() {
        if (mMediaRecorder != null) {
            var maxAmplitude = mMediaRecorder?.maxAmplitude ?: 0
            var level = 9* 3 * maxAmplitude / (32768*2) + 1
            if (mVolumeListener != null) {
                mVolumeListener?.invoke(level)
            }
            getHandler()?.removeMessages(MSG_UPDATE_VOLUME)
            getHandler()?.sendEmptyMessageDelayed(MSG_UPDATE_VOLUME, 200)
        }
    }

    private fun getHandler(): Handler? {
        if (mUiHandler == null) {
            mUiHandler = object : Handler(Looper.getMainLooper()) {
                override fun handleMessage(msg: Message) {
                    if (msg.what == MSG_UPDATE_VOLUME) {
                        updateVolume()
                    }
                }
            }
        }
        return mUiHandler;
    }

    class Builder internal constructor() {
        internal var mParams = MyMediaRecorder()

        fun setAudioSource(audioSource: Int): Builder {
            mParams.audioSource = audioSource
            return this
        }

        fun setOutputFormat(outputFormat: Int): Builder {
            mParams.outputFormat = outputFormat
            return this
        }

        fun setAudioEncoder(audioEncoder: Int): Builder {
            mParams.audioEncoder = audioEncoder
            return this
        }

        fun setAudioChannel(audioChannel: Int): Builder {
            mParams.audioChannel = audioChannel
            return this
        }

        fun setAudioSamplingRate(audioSamplingRate: Int): Builder {
            mParams.audioSamplingRate = audioSamplingRate
            return this
        }

        fun setAudioEncodingBitRate(audioEncodingBitRate: Int): Builder {
            mParams.audioEncodingBitRate = audioEncodingBitRate
            return this
        }

        fun build(): MyMediaRecorder {
            return mParams
        }

    }

    companion object {
        val TAG = "MyMediaRecorder"

        fun newBuilder(): Builder {
            return Builder()
        }
    }

}
